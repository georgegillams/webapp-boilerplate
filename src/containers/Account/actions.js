import { createRoutine } from 'redux-saga-routines';

import { LOAD_SHOWCASES } from './constants';

export const loadShowcases = createRoutine(LOAD_SHOWCASES);
