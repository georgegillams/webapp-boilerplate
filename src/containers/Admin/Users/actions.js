import { createRoutine } from 'redux-saga-routines';

import { LOAD, REMOVE, REQUEST_MAGIC_LINK } from './constants';

export const load = createRoutine(LOAD);
export const remove = createRoutine(REMOVE);
export const requestMagicLink = createRoutine(REQUEST_MAGIC_LINK);
