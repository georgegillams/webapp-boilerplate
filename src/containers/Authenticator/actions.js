import { createRoutine } from 'redux-saga-routines';

import { LOAD_AUTH, SET_COOKIES_ALLOWED } from './constants';

export const loadAuth = createRoutine(LOAD_AUTH);
export const setCookiesAllowed = createRoutine(SET_COOKIES_ALLOWED);
