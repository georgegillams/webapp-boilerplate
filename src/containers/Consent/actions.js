import { createRoutine } from 'redux-saga-routines';

import { CONSENT, DEFER_CONSENT, RE_PROMPT_CONSENT, RESET_CONSENT } from './constants';

export const consent = createRoutine(CONSENT);
export const deferConsent = createRoutine(DEFER_CONSENT);
export const rePromptConsent = createRoutine(RE_PROMPT_CONSENT);
export const resetConsent = createRoutine(RESET_CONSENT);
