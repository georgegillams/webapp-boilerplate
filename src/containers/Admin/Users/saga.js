import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'utils/request';
import apiStructure from 'helpers/apiStructure';

import { selectState } from './selectors';
import { load, remove, requestMagicLink } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadUsers.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.users));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doRemove() {
  const currentState = yield select(selectState());
  const { userToRemove } = currentState;
  const requestURL = apiStructure.deleteUser.fullPath;

  try {
    yield put(remove.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(userToRemove),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(remove.failure(result));
    } else {
      yield put(remove.success(result));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(remove.failure(err));
  } finally {
    yield put(remove.fulfill());
  }
}

export function* doRequestMagicLink() {
  const currentState = yield select(selectState());
  const { userToLogIn } = currentState;
  const requestURL = apiStructure.requestMagicLink.fullPath;

  try {
    yield put(requestMagicLink.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ email: userToLogIn.email, divertToAdmin: true }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(requestMagicLink.failure(result));
    } else {
      yield put(requestMagicLink.success(result));
    }
  } catch (err) {
    yield put(requestMagicLink.failure(err));
  } finally {
    yield put(requestMagicLink.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(remove.TRIGGER, doRemove);
  yield takeLatest(requestMagicLink.TRIGGER, doRequestMagicLink);
}
