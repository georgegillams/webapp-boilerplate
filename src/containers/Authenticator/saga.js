import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'utils/request';
import apiStructure from 'helpers/apiStructure';

import { loadAuth } from './actions';

export function* doLoadAuth() {
  // const currentState = yield select(selectState());
  // console.log(`currentState`, currentState);

  const requestURL = apiStructure.loadAuth.fullPath;

  try {
    yield put(loadAuth.request());

    const result = yield call(request, requestURL);

    yield put(loadAuth.success(result));
  } catch (err) {
    yield put(loadAuth.failure(err));
  } finally {
    yield put(loadAuth.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadAuth.TRIGGER, doLoadAuth);
}
