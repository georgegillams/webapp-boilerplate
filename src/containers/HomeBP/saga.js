import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'utils/request';

import { loadShowcases } from './actions';

export function* loadShowcasesRequest() {
  const requestURL = 'https://us-central1-react-next-boilerplate-cda8b.cloudfunctions.net/getShowcasesData';

  try {
    yield put(loadShowcases.request());

    const result = yield call(request, requestURL);

    yield put(loadShowcases.success(result));
  } catch (err) {
    yield put(loadShowcases.failure(err));
  } finally {
    yield put(loadShowcases.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadShowcases.TRIGGER, loadShowcasesRequest);
}
