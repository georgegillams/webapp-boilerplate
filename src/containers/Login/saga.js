import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'utils/request';
import apiStructure from 'helpers/apiStructure';
import { selectState } from './selectors';

import { login } from './actions';

export function* doLogin() {
  const currentState = yield select(selectState());
  const { credentials } = currentState;

  const requestURL = apiStructure.requestMagicLink.fullPath;

  try {
    yield put(login.request());

    // TODO should pass loginRedirect to the API
    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ ...credentials }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    yield put(login.success(result));
  } catch (err) {
    yield put(login.failure(err));
  } finally {
    yield put(login.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(login.TRIGGER, doLogin);
}
