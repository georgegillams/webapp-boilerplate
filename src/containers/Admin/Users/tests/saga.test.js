import { put, takeLatest } from 'redux-saga/effects';

import { load, remove, requestMagicLink } from '../actions';

import saga, { doLoad, doRemove, doRequestMagicLink } from '../saga';

describe('AdminUsers saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should load on load TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(load.TRIGGER, doLoad));
  });

  it('Should remove on remove TRIGGER', () => {
    mainSaga.next();
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(remove.TRIGGER, doRemove));
  });

  it('Should request on request TRIGGER', () => {
    mainSaga.next();
    mainSaga.next();
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(requestMagicLink.TRIGGER, doRequestMagicLink));
  });

  describe('admin user actions', () => {
    let loadGenerator;
    let removeGenerator;
    let requestGenerator;

    const loadResponse = {
      users: ['user1', 'user2'],
      status: 200,
    };
    const removeResponse = {
      success: 'User removed',
      status: 200,
    };
    const requestResponse = {
      success: 'Magic link sent',
      status: 200,
    };

    beforeEach(() => {
      loadGenerator = doLoad();
      removeGenerator = doRemove();
      requestGenerator = doRequestMagicLink();

      const selectLoadDescriptor = loadGenerator.next().value;
      const selectRemoveDescriptor = removeGenerator.next().value;
      const selectRequestDescriptor = requestGenerator.next().value;
      expect(selectLoadDescriptor + selectRemoveDescriptor + selectRequestDescriptor).toMatchSnapshot();
    });

    it('Should call load.success on successful API call', () => {
      const response = loadResponse;
      loadGenerator.next();
      const putSuccess = loadGenerator.next(response).value;
      loadGenerator.next();

      expect(putSuccess).toEqual(put(load.success(response.users)));
    });

    it('Should call verify.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      loadGenerator.next();
      const putFailure = loadGenerator.next(response).value;
      loadGenerator.next();

      expect(putFailure).toEqual(put(load.failure(response)));
    });

    it('Should call verify.failure if an exception occurs', () => {
      const response = new Error('Some error');
      const putFailure = loadGenerator.throw(response).value;

      expect(putFailure).toEqual(put(load.failure(response)));
    });

    it('Should call remove.success on successful API call', () => {
      const response = removeResponse;
      removeGenerator.next({ userToRemove: { id: 'u1' } });
      removeGenerator.next();
      const putSuccess = removeGenerator.next(response).value;
      removeGenerator.next();

      expect(putSuccess).toEqual(put(remove.success(response)));
    });

    it('Should call load.trigger on successful API call', () => {
      const response = removeResponse;
      removeGenerator.next({ userToRemove: { id: 'u1' } });
      removeGenerator.next();
      removeGenerator.next(response);
      const putSuccess = removeGenerator.next().value;
      removeGenerator.next();

      expect(putSuccess).toEqual(put(load.trigger()));
    });

    it('Should call remove.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      removeGenerator.next({ userToRemove: { id: 'u1' } });
      removeGenerator.next();
      const putFailure = removeGenerator.next(response).value;
      removeGenerator.next(response);

      expect(putFailure).toEqual(put(remove.failure(response)));
    });

    it('Should call remove.failure if an exception occurs', () => {
      const response = new Error('Some error');
      removeGenerator.next({ userToRemove: { id: 'u1' } });
      const putFailure = removeGenerator.throw(response).value;

      expect(putFailure).toEqual(put(remove.failure(response)));
    });

    it('Should call requestMagicLink.success on successful API call', () => {
      const response = requestResponse;
      requestGenerator.next({ userToLogIn: { id: 'u1' } });
      requestGenerator.next();
      const putSuccess = requestGenerator.next(response).value;
      requestGenerator.next();

      expect(putSuccess).toEqual(put(requestMagicLink.success(response)));
    });

    it('Should call requestMagicLink.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      requestGenerator.next({ userToLogIn: { id: 'u1' } });
      requestGenerator.next();
      const putFailure = requestGenerator.next(response).value;
      requestGenerator.next(response);

      expect(putFailure).toEqual(put(requestMagicLink.failure(response)));
    });

    it('Should call requestMagicLink.failure if an exception occurs', () => {
      const response = new Error('Some error');
      requestGenerator.next({ userToLogIn: { id: 'u1' } });
      const putFailure = requestGenerator.throw(response).value;

      expect(putFailure).toEqual(put(requestMagicLink.failure(response)));
    });
  });
});
