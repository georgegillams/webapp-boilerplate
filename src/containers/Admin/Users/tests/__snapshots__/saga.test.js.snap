// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminUsers saga admin user actions Should call load.success on successful API call 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call load.trigger on successful API call 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call remove.failure if API returns error 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call remove.failure if an exception occurs 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call remove.success on successful API call 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call requestMagicLink.failure if API returns error 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call requestMagicLink.failure if an exception occurs 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call requestMagicLink.success on successful API call 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call verify.failure if API returns error 1`] = `"[object Object][object Object][object Object]"`;

exports[`AdminUsers saga admin user actions Should call verify.failure if an exception occurs 1`] = `"[object Object][object Object][object Object]"`;
